### ISSUE TITLE
[Feature]: Support setting unix socket listener file mode

### ISSUE THREAD

**Issue by Tristan971** [oauth2-proxy/oauth2-proxy#2536]
### Motivation

Typically, it is best to avoid running programs as root.

Unfortunately, the current unix socket binding logic sets the socket's file mode to 0644 (default behaviour of the Go API for net.Listen, not configurable).

So when using a reverse proxy and oauth2-proxy on the side, it is very impractical to have our reverse proxy call oauth2-proxy (unless our reverse proxy runs as root, which it preferrably shouldn't need to).

It would therefore be great if it was possible to have oauth2-proxy bind sockets using a specific unix file mode.

### Possible solution

There are 2 ways to set socket permissions in Go, as per https://github.com/golang/go/issues/11822:
1. change your process' default filemode permissions
2. invoke os.Chmod on the socket file just created

Obviously the first option is quite undesirable, and the second one is simple enough, so let's go with chmod.

When it comes to the configuration format, there are 2 approaches:
1. Make BindAddress a struct, with an optional UnixSocketFileMode field
2. Append file attribute arguments after the end of the unix path, in comma-separated key-value form for example[1] eg: `unix:///path/to/file.sock,mode=0775`

[1] Similar a bunch of other software, most notably docker compose short-form volumes and Varnish

Again, the first option seems quite undesirable (both in terms of complexity for a niche use-case, and in terms of being backwards-incompatible with both the legacy and alpha configuration formats).

Therefore I suggest option 2. If necessary/asked in the future, it also nicely accounts for changing the user/group owning the socket, and any such other file attributes related needs.

A final solution could be to get https://github.com/oauth2-proxy/oauth2-proxy/pull/1985 merged, but that doesn't look very promising at the moment, and makes systemd a hard requirement for it which might not be desirable/doable with containers.

### Provider

None

**Comment by isodude** [comment]
Systemd is no way a hard requirement for #1985, just as long as you set LISTEN_FDS=1 LISTEN_PID=2 and --http-address=fd:3 you should be fine (provided there an actual listening socket on fd 3).

`systemd-socket-activate` is a standalone program that you can prepend to oauth2-proxy also. It's quite easy to replicate it also.
`systemd-socket-activate -l /tmp/socket <oauth2-proxy>` would start the process when there's traffic on the socket. What's funny is that systemd-socket-activate does not have support for changing mode :-).

Hence I do think that adding mode= to http-address would make sense for unix addresses.

For my part it's pretty much solved with the systemd support and a systemd.socket, as long as the socket has mode 0660 and group www-data, nginx have no problem accessing it. The problem with the patch seems to be very much some odd problem in Go when listening on new sockets, but it's mostly causing problems in tests.

**Comment by Tristan971** [comment]
> systemd-socket-activate is a standalone program that you can prepend to oauth2-proxy also.

Oh, neat!

> What's funny is that systemd-socket-activate does not have support for changing mode :-).

Oh come on... sigh... :joy: 

Well, I guess both patchis still make sense then... sigh...

**Comment by Tristan971** [comment]
Well, I never knew. That certainly solves the issue for me (have to check that it works nicely but I see no reason why it wouldn’t :)

**Comment by Tristan971** [comment]
For others maybe looking into this problem for similar reasons (unprivileged reverse proxy user calling oauth2-proxy's socket), abstract unix sockets happen to do the job fine:

```toml
# oauth2-proxy.toml

http_address = "unix://@oauth2-proxy"
# ...
skip_auth_routes = ["^/ping", "^/metrics"]
# ...
```

```bash
$ sudo -u unprivileged-user \
  curl -I --abstract-unix-socket oauth2-proxy http://foo/ping

HTTP/1.1 200 OK
Date: Wed, 06 Mar 2024 02:47:33 GMT
Content-Length: 2
Content-Type: text/plain; charset=utf-8
```

So, assuming your reverse proxy can use these, you then don't need to worry about permissions or filepaths

**Comment by isodude** [comment]
Isn't that the same though as allowing it to the whole system?
You might want to investigate `PrivateIPC` etc.

**Comment by Tristan971** [comment]
> Isn't that the same though as allowing it to the whole system?

It is allowing the whole system to send requests to oauth2-proxy, yes. Which is not entirely ideal, but a better tradeoff still than allowing the reverse proxy to access all the files that oauth2-proxy can.

I'd also note that when using TCP listeners, the problem is the same, unless you dedicated network namespaces to all your processes etc, so it is more like "same semantics as TCP listeners, without reserving a port".

**Comment by isodude** [comment]
Correct, same same, but different :) Easier to userspace that PrivateNetwork for random processes also, but no iptables capability.

I have been playing about with PrivateIPC in systemd, but it's very bothersome to use when you have this scenario.

It should be noted that the systemd-proxyd way actually solves it by having root forward the packets instead. With the extra layer of a service that is though. That way neither nginx nor oauth2-proxy have any elevated privileges.
